src/Microsoft.Framework.Runtime.Hosting/Json/JsonDeserializer.cs

```diff
@@ -0,0 +1,179 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+
+namespace Microsoft.Framework.Runtime.Json
+{
+    internal static class JsonDeserializer
+    {
+        public static JsonValue Deserialize(TextReader reader)
+        {
+            if (reader == null)
+            {
+                throw new ArgumentNullException(nameof(reader));
+            }
+
+            var buffer = new JsonBuffer(reader);
+
+            var result = DeserializeInternal(buffer.Read(), buffer);
+
+            // There are still unprocessed char. The parsing is not finished. Error happened.
+            var nextToken = buffer.Read();
+            if (nextToken.Type != JsonTokenType.EOL)
+            {
+                throw new JsonDeserializerException(
+                    JsonDeserializerResource.Format_UnfinishedJSON(nextToken.Value),
+                    nextToken);
+            }
+
+            return result;
+        }
+
+        private static JsonValue DeserializeInternal(JsonToken next, JsonBuffer buffer)
```

Comment:
I agree with @Eilon . It is simpler to keep the `JsonDeserializer` a static class. Plus `TestReader` is a disposable class. I won't want to keep a handle of it.